src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,426 @@
+   pos=1
+   for i in 1:length(L)-1
+      pos += L[i][1]*L[i][2]
+      # block above diagonal
+      z = idN
+      for j in 1:L[i][1]
+         idx = pos-L[i][1]+j-1
+         z[idx,idx+L[i+1][1]] = 1
+      end
+      push!(listgens,z)
+      # block below diagonal
+      z = idN
+      for j in 1:L[i][1]
+         idx = pos+j-1
+         z[idx,idx-L[i][1]]=1
+      end
+      push!(listgens,z)
+   end
+
+   # cardinality
+   res = prod([_GL_order(l[2],F) for l in L])
+   exp = fmpz(0)
+   for i in 1:length(L)-1, j in i+1:length(L)
+      exp += L[i][1]*L[i][2]*L[j][2]
+   end
+   exp *= 2
+   exp += sum([(L[i][1]-1)*L[i][2]^2 for i in 1:length(L)])
+   res *= order(F)^exp
+
+   return listgens, res
+end
+
+
+# V = vector of integers of the dimensions of Jordan blocks
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+# does the same as above, but every entry is replaced by the corresponding block matrix in the field F/(f)
+function _centr_block_unipotent(f::PolyElem, F::FinField, V::AbstractVector{Int}; isSL=false)
+   d = degree(f)
+   d>1 || return _centr_unipotent(F,V; isSL=isSL)
+   n = sum(V)*d
+   C = companion_matrix(f)
+   idN = identity_matrix(F,n)
+   idD = identity_matrix(F,d)
+
+   # L = multiset(V)
+   L=[[V[1],1]]
+   for i in 2:length(V)
+      if V[i]==L[length(L)][1]
+         L[length(L)][2] += 1
+      else
+         push!(L, [V[i],1])
+      end
+   end
+   listgens = MatElem[]
+
+   # generators for GL + internal diagonal blocks
+   pos=1
+   for l in L
+      v_g = isSL ? _gens_for_SL_matrix(f,l[2],F; D=l[1]) : _gens_for_GL_matrix(f,l[2],F; D=l[1])
+      for x in v_g
+         z = insert_block(idN,x,pos,pos)
+         push!(listgens,z)
+      end
+      if l[1]>1
+         for i in 1:l[1]-1
+            c = idD
+            for j in 1:degree(F)*d
+               z = identity_matrix(F,l[1]*d)
+               for k in 0:l[1]-i-1 insert_block!(z,c,d*k+1 ,d*(k+i)+1) end
+               z = insert_block(idN,z,pos,pos)
+               c *= C           # every time, the block C^(j-1) is inserted
+               push!(listgens,z)
+            end
+         end
+      end
+      pos += l[1]*l[2]*d
+   end
+
+   # external diagonal blocks
+   pos=1
+   for i in 1:length(L)-1
+      pos += L[i][1]*L[i][2]*d
+      # block above diagonal
+      z = idN
+      for j in 1:L[i][1]*d
+         idx = pos-L[i][1]*d+j-1
+         z[idx,idx+L[i+1][1]*d]=1
+      end
+      push!(listgens,z)
+      # block below diagonal
+      z = idN
+      for j in 1:L[i][1]*d
+         idx = pos+j-1
+         z[idx,idx-L[i][1]*d]=1
+      end
+      push!(listgens,z)
+   end
+
+   # cardinality
+   res = prod([_GL_order(l[2],order(F)^degree(f)) for l in L])
+   exp = fmpz(0)
+   for i in 1:length(L)-1, j in i+1:length(L)  exp += L[i][1]*L[i][2]*L[j][2] end
+   exp *= 2
+   exp += sum([(L[i][1]-1)*L[i][2]^2 for i in 1:length(L)])
+   exp *= degree(f)
+   res *= order(F)^exp
+
+   return listgens, res
+end
+
+# returns the list of generators
```

Comment:
It also returns two more values `res` and `cbm`, can you please also describe them in this comment, even if briefly?

Reply:
Done it. I just moved the change of basis from the function `centralizer` to the function `_centralizer_GL` and `_centralizer_SL`, since I just realized that it is a cleaner way to proceed.